modin/experimental/core/execution/dask/implementations/pandas_on_dask/io/io.py

```diff
@@ -0,0 +1,257 @@
+        parse_dates : list or dict, optional
+            The behavior is as follows:
+
+            - List of column names to parse as dates.
+            - Dict of `{column_name: format string}` where format string is
+              strftime compatible in case of parsing string times, or is one of
+              (D, s, ns, ms, us) in case of parsing integer timestamps.
+            - Dict of `{column_name: arg dict}`, where the arg dict corresponds
+              to the keyword arguments of ``pandas.to_datetime``.
+              Especially useful with databases without native Datetime support,
+              such as SQLite.
+        columns : list, optional
+            List of column names to select from SQL table (only used when reading a
+            table).
+        chunksize : int, optional
+            If specified, return an iterator where `chunksize` is the number of rows
+            to include in each chunk.
+        partition_column : str, optional
+            Column name used for data partitioning between the workers
+            (MUST be an INTEGER column).
+        lower_bound : int, optional
+            The minimum value to be requested from the `partition_column`.
+        upper_bound : int, optional
+            The maximum value to be requested from the `partition_column`.
+        max_sessions : int, optional
+            The maximum number of simultaneous connections allowed to use.
+
+        Returns
+        -------
+        BaseQueryCompiler
+            A new query compiler with imported data for further processing.
+        """
+        from modin.experimental.core.io.sql.utils import (
+            is_distributed,
+            get_query_info,
+            read_sql_with_offset,
+        )
+
+        if not is_distributed(partition_column, lower_bound, upper_bound):
+            warnings.warn("Defaulting to Modin core implementation")
+            return PandasOnDaskIO.read_sql(
+                sql,
+                con,
+                index_col,
+                coerce_float=coerce_float,
+                params=params,
+                parse_dates=parse_dates,
+                columns=columns,
+                chunksize=chunksize,
+            )
+        #  starts the distributed alternative
+        cols_names, query = get_query_info(sql, con, partition_column)
+        num_parts = min(NPartitions.get(), max_sessions if max_sessions else 1)
+        num_splits = min(len(cols_names), num_parts)
+        diff = (upper_bound - lower_bound) + 1
+        min_size = diff // num_parts
+        rest = diff % num_parts
+        partition_ids = []
+        index_ids = []
+        end = lower_bound - 1
+        for part in range(num_parts):
+            if rest:
+                size = min_size + 1
+                rest -= 1
+            else:
+                size = min_size
+            start = end + 1
+            end = start + size - 1
+            partition_id = DaskWrapper.deploy(
```

Comment:
can we get rid of copy-pasting here as well? It feels like the only thing that has changed from Ray's implementation is that it now uses `DaskWrapper` and `DaskPartitions` classes. Can we separate the base implementation and parametrize it with the engine-specific classes?

Reply:
the same question for `to_pickle_distributed`